new guide on profiling workers (#2330)
cloudflare/cloudflare-docs: products/workers/src/content/learning/profiling-workers.md

---
order: 4
pcx-content-type: concept
---

# Profiling Workers

For long-running or complex Workers, you need to understand where your code is spending the most time.

> That's why we're launching a beta version of wrangler inspect. We're beginning to integrate the Workers platform with the Chrome Developer Tools, and starting today, you can use `wrangler dev --inspect` to get a CPU profile of your Worker.

Reviewer: ```suggestion
That's why we're launching a beta version of `wrangler inspect`. We're beginning to integrate the Workers platform with the Chrome Developer Tools, and starting today, you can use `wrangler dev --inspect` to get a CPU profile of your Worker.
```

To do this, first, make sure you're on Wrangler version v1.19.3 or later.

## Profiling an example project

> Let's create a brand new Worker with:

Reviewer 2: @deadlypants1973 will hunt down all `Let's` and `we`s – style guide

Author: Gonna be my hardest habit to break 😭

```sh

> wrangler generate my-worker

Reviewer: ```suggestion
$ wrangler generate my-worker
```

```

Open `index.js` in your IDE and replace the contents with:

```javascript
addEventListener("fetch", (event) => {
  event.respondWith(handleRequest(event.request));
});

/**
 * Respond with hello worker text
 * @param {Request} request
 */
const timer = (ms) => new Promise((res) => setTimeout(res, ms));

async function sleepBetween() {
  let result = 0;

  for (let i = 0; i < 100; i++) {
    result += i;
    await timer(100);
  }

  return result;
}

async function handleRequest(request) {
  await sleepBetween();

  return new Response("Hello worker!", {
    headers: { "content-type": "text/plain" },
  });
}
```

This is a basic example where our request handler calls an async function `sleepBetween`. The function iterates from 0 to 100, but after each iteration, it calls `timer(100)`, which will sleep for 100 milliseconds. Now we have a nice slow function we can use to test out the profiler.

### Profiling our Worker

With this function saved, go to your terminal and run `wrangler dev --inspect`. You'll see some new instructions for configuring Chrome DevTools.

```sh

> ☁  my-worker [master] ⚡  wrangler dev --inspect

Reviewer: Prefix sh code with `$`

💁  watching "./"
🕵️  Open chrome://inspect, click 'Configure', and add localhost:9230
👂  Listening on http://127.0.0.1:8787
```

Open a new tab in Chrome, type in `chrome://inspect` and hit enter. Click on "Configure" and add our new localhost port, `localhost:9230`. Now you should see the Wrangler inspect process showing up under "Remote Target."

![Chrome Inspect](./media/chrome-inspect.png)

Clicking inspect opens up the DevTools. Currently, Wrangler only supports the Console, Sources and Profiler tabs.

### Capturing a CPU profile

With the DevTools open, click the Profile tab and then click "start." Now open a new tab with your Worker running locally `http://127.0.0.1:8787`. Accessing the Worker causes it to run again, this time captured by the DevTools. When it finishes loading, go back to our DevTools window and click "stop."

There are three ways to view CPU profile data: tree (top down), heavy (bottom up) and chart. If you click on "Chart," it brings us to a flame chart where we can see every function call and how long it took.

> My advice for finding slow functions is to look for the last long function in a stack. Since each function's total run time is determined by the runtime of all the functions it calls. We want to find the one with the longest self runtime, or the last big one before it finishes or splits into smaller calls. For more information on using the DevTools, check out their [official docs](https://developer.chrome.com/docs/devtools/).

Reviewer: ```suggestion
To find slow functions, look for the last long function in a stack. Since each function's total run time is determined by the runtime of all the functions it calls. We want to find the one with the longest self runtime, or the last big one before it finishes or splits into smaller calls. For more information on using the DevTools, check out their [official docs](https://developer.chrome.com/docs/devtools/).
```

---
order: 4
pcx-content-type: concept
---

# Profiling Workers

For long-running or complex Workers, you need to understand where your code is spending the most time.

That's why we're launching a beta version of `wrangler inspect`. We're beginning to integrate the Workers platform with the Chrome Developer Tools, and starting today, you can use `wrangler dev --inspect` to get a CPU profile of your Worker.

<StreamVideo id="f11809a382160334e9be9a2aedf13d1d" />

> To do this, first, make sure you're on Wrangler version v1.19.3 or later.

Reviewer: can we link to the updating wrangler page: https://developers.cloudflare.com/workers/cli-wrangler/install-update

## Profiling an example project

> You can reate a brand new Worker with:

Reviewer: ```suggestion
You can create a new Workers project with the `wrangler generate` command:
```

```sh
$ wrangler generate my-worker
```

Open `index.js` in your IDE and replace the contents with:

```javascript
addEventListener("fetch", (event) => {
  event.respondWith(handleRequest(event.request));
});

/**
 * Respond with hello worker text
 * @param {Request} request
 */
const timer = (ms) => new Promise((res) => setTimeout(res, ms));

async function sleepBetween() {
  let result = 0;

  for (let i = 0; i < 100; i++) {
    result += i;
    await timer(100);
  }

  return result;
}

async function handleRequest(request) {
  await sleepBetween();

  return new Response("Hello worker!", {
    headers: { "content-type": "text/plain" },
  });
}
```

> This is a basic example where our request handler calls an async function `sleepBetween`. The function iterates from 0 to 100, but after each iteration, it calls `timer(100)`, which will sleep for 100 milliseconds. Now there is a nice slow function you can use to test out the profiler.

Reviewer: ```suggestion
This is a basic example where our request handler calls an async function `sleepBetween`. The function iterates from 0 to 100, but after each iteration, it calls `timer(100)`, which will sleep for 100 milliseconds. This function will simulate slow code, which you can use to test out the profiler.
```

### Profiling our Worker

With this function saved, go to your terminal and run `wrangler dev --inspect`. You'll see some new instructions for configuring Chrome DevTools.

```sh

> $ ☁  my-worker [master] ⚡  wrangler dev --inspect

Reviewer: Can we remove emoji on this line and just do `$ wrangler dev --inspect`

💁  watching "./"
🕵️  Open chrome://inspect, click 'Configure', and add localhost:9230
👂  Listening on http://127.0.0.1:8787
```

> Open a new tab in Chrome, type in `chrome://inspect` and hit enter. Click on "Configure" and add our new localhost port, `localhost:9230`. Now you should see the Wrangler inspect process showing up under "Remote Target."

Reviewer: ```suggestion
Open a new tab in Chrome, type in `chrome://inspect` and hit enter. Click on "Configure" and add `localhost:9230`. Now you should see the Wrangler inspect process showing up under "Remote Target."
```

![Chrome Inspect](./media/chrome-inspect.png)

> Clicking inspect opens up the DevTools. Currently, Wrangler only supports the Console, Sources and Profiler tabs.

Reviewer: ```suggestion
Clicking inspect opens up the DevTools. 

<Aside>

Currently, Wrangler only supports the Console, Sources and Profiler tabs.

</Aside>
```

### Capturing a CPU profile

With the DevTools open, click the Profile tab and then click "start." Now open a new tab with your Worker running locally `http://127.0.0.1:8787`. Accessing the Worker causes it to run again, this time captured by the DevTools. When it finishes loading, go back to our DevTools window and click "stop."

> There are three ways to view CPU profile data: tree (top down), heavy (bottom up) and chart. If you click on "Chart," it brings us to a flame chart where you can see every function call and how long it took.

Reviewer: I always get our style guide mixed up here, unsure if it should be "it will bring you" or "you will see" or what the right subject/tense here. @lukeed help!!!

> There are three ways to view CPU profile data: tree (top down), heavy (bottom up) and chart. If you click on "Chart," it brings us to a flame chart where you can see every function call and how long it took.

Reviewer 2: The `tree (top down), heavy (bottom up) and chart.` part should be expanded a little bit more, and if `tree` is referencing the display name, I think it needs to be quoted? I'd format it something like this:

```
There are three ways to view the CPU profiling data: 

1. "Tree" – a top-down view of __ for ___;
2. "Heavy" – a bottom-up view of __ for __; and
3. "Chart" – a flame chart which will show all function calls and how much time was spent in each function execution.

To find slow functions, enter the "Chart" view and search for the last long function in a stack (ideally a screenshot). // snip
```

Suggestion for the style-guide tense changes:

```suggestion
There are three ways to view CPU profile data: tree (top down), heavy (bottom up) and chart. If you click on "Chart," it brings us to a flame chart which will show all function calls and how much time was spent in each function execution.
```

> To find slow functions, look for the last long function in a stack. Since each function's total run time is determined by the runtime of all the functions it calls. We want to find the one with the longest self runtime, or the last big one before it finishes or splits into smaller calls. For more information on using the DevTools, check out their [official docs](https://developer.chrome.com/docs/devtools/).

Reviewer: ```suggestion
To find slow functions, look for the last long function in a stack, sSince each function's total run time is determined by the runtime of all the functions it calls. We want to find the one with the longest self runtime, or the last big one before it finishes or splits into smaller calls. For more information on using the DevTools, check out Chrome's [official DevTools docs](https://developer.chrome.com/docs/devtools/).
```

Reviewer 2: Need to rework "We" out of it tho 🙃 

Also, isn't this paragraph just repeating itself? First two sentences say the same thing. Repetition stops at `For more information`

There are three ways to view the CPU profiling data:

1. **Tree** - a top-down view of all functions called. Starting with top-level functions and nesting subsequent calls under each one.
2. **Heavy** - a bottom-up view of all functions called. Starting with the final functions in the stack and nesting parent calls under each one.
3. **Chart** - a flame chart that will show all function calls and how much time was spent in each function execution.

To find slow functions, enter the **Chart** view and search for the last long function in a stack.

![CPU Flame chart](./media/devtools-chart.png)

A function's total runtime is determined by the runtime of all the functions it calls. You will want to find the function with the longest "Self Time", or the last big runtime before it finishes or splits into smaller calls. For more information on using the DevTools, refer to the [official Chrome documentation](https://developer.chrome.com/docs/devtools/).

In the example image above, `sleepBetween` has the longest runtime.

> You can also arrive at this conclusion through the **Heavy** view, which allows you to sort by "Self Time" or "Total Time". When you sort by "Self Time" and ignore any items in parenthesis, you can identify the function with the slowest runtime.

Reviewer 2: ```suggestion
You can also arrive at this conclusion through the **Heavy** view, which allows you to sort by "Self Time" or "Total Time". When you sort by "Self Time" and ignore any items in parenthesis, you can identify the functions with the longest execution time.
```

---
order: 4
pcx-content-type: concept
---

# Profiling Workers

> For long-running or complex Workers, you need to understand where your code is spending the most time.
> 
> That's why we're launching a beta version of `wrangler inspect`. We're beginning to integrate the Workers platform with the Chrome Developer Tools, and starting today, you can use `wrangler dev --inspect` to get a CPU profile of your Worker.

Reviewer 3: ```suggestion
For long-running or complex Workers, `wrangler inspect`(beta) will help you need to understand where your code is spending the most time.

As part of the Workers platform's integration with Chrome Developer Tools, you can use `wrangler dev --inspect` to get a CPU profile of your Worker.
```

<StreamVideo id="f11809a382160334e9be9a2aedf13d1d" />

> To do this, first, make sure you're on Wrangler version v1.19.3 or later. Here is a guide to [upgrading Wrangler](https://developers.cloudflare.com/workers/cli-wrangler/install-update)

Reviewer 3: ```suggestion
To run `wrangler inspect`, first, make sure you are on Wrangler version v1.19.3 or later. If you need to update your version, refer to the guide on [upgrading Wrangler](https://developers.cloudflare.com/workers/cli-wrangler/install-update).
```

## Profiling an example project

> You can create a new Workers project with the `wrangler generate` command:

Reviewer 3: ```suggestion
Create a new Workers project with the `wrangler generate` command:
```

```sh
$ wrangler generate my-worker
```

> Open `index.js` in your IDE and replace the contents with:

Reviewer 3: ```suggestion
Open the `index.js` file in your project's directory and replace the contents with:
```

```javascript
addEventListener("fetch", (event) => {
  event.respondWith(handleRequest(event.request));
});

/**
 * Respond with hello worker text
 * @param {Request} request

>  */

Reviewer 2: this is out of place

const timer = (ms) => new Promise((res) => setTimeout(res, ms));

async function sleepBetween() {
  let result = 0;

  for (let i = 0; i < 100; i++) {
    result += i;
    await timer(100);
  }

  return result;
}

async function handleRequest(request) {
  await sleepBetween();

  return new Response("Hello worker!", {
    headers: { "content-type": "text/plain" },
  });
}
```

> This is a basic example where our request handler calls an async function `sleepBetween`. The function iterates from 0 to 100, but after each iteration, it calls `timer(100)`, which will sleep for 100 milliseconds. This function will simulate slow code, which you can use to test out the profiler.

Reviewer 3: ```suggestion
This is a basic example where the Worker request handler calls an async function `sleepBetween`. The function iterates from 0 to 100, but after each iteration, it calls `timer(100)`, which will sleep for 100 milliseconds. This function will simulate slow code, which you can use to test out the profiler.
```

### Profiling our Worker

> With this function saved, go to your terminal and run `wrangler dev --inspect`. You'll see some new instructions for configuring Chrome DevTools.

Reviewer 3: ```suggestion
With this function saved, go to your terminal and run `wrangler dev --inspect`. You will see additional instructions for configuring Chrome DevTools.
```

```sh
$ wrangler dev --inspect
💁  watching "./"
🕵️  Open chrome://inspect, click 'Configure', and add localhost:9230
👂  Listening on http://127.0.0.1:8787
```

> Open a new tab in Chrome, type in `chrome://inspect` and hit enter. Click on "Configure" and add `localhost:9230`. Now you should see the Wrangler inspect process showing up under "Remote Target."

Reviewer 3: ```suggestion
Open a new tab in Chrome, type in `chrome://inspect` and hit enter. Select **Configure** and add `localhost:9230`. The Wrangler inspect process will then show up under **Remote Target**.
```

![Chrome Inspect](./media/chrome-inspect.png)

> Clicking inspect opens up the DevTools.

Reviewer 3: ```suggestion
Selecting **Inspect** opens up the DevTools.
```

> <Aside>

Reviewer 3: ```suggestion
<Aside type="note">
```

Currently, Wrangler only supports the Console, Sources and Profiler tabs.

</Aside>

### Capturing a CPU profile

> With the DevTools open, click the Profile tab and then click "start." Now open a new tab with your Worker running locally `http://127.0.0.1:8787`. Accessing the Worker causes it to run again, this time captured by the DevTools. When it finishes loading, go back to our DevTools window and click "stop."

Reviewer 3: ```suggestion
With the DevTools open, select the **Profile** tab > **Start**. Open a new tab with your Worker running locally `http://127.0.0.1:8787`. Accessing the Worker causes it to run again, this time captured by the DevTools. When it finishes loading, go back to our DevTools window and select **Stop**.
```

> With the DevTools open, select the **Profile** tab > **Start**. Open a new tab with your Worker running locally `http://127.0.0.1:8787`. Accessing the Worker causes it to run again, this time captured by the DevTools. When it finishes loading, go back to our DevTools window and select **Stop**.

Reviewer 2: ```suggestion
With the DevTools open, select the **Profile** tab > **Start**. Open a new tab with your Worker running locally `http://127.0.0.1:8787`. Visiting this address will invoke the Worker again, allowing DevTools to capture execution information. When it finishes loading, go back to our DevTools window and select **Stop**.
```

There are three ways to view the CPU profiling data:

> 1. "Tree" - a top-down view of all functions called. Starting with top level functions and nesting subsequent calls under each one.
> 2. "Heavy" - a bottom up view of all functions called. Starting with the final functions in the stack and nesting parent calls under each one.
> 3. "Chart" - a flame chart which will show all function calls and how much time was spent in each function execution.

Reviewer 3: ```suggestion
1. **Tree** - a top-down view of all functions called. Starting with top-level functions and nesting subsequent calls under each one.
2. **Heavy** - a bottom-up view of all functions called. Starting with the final functions in the stack and nesting parent calls under each one.
3. **Chart** - a flame chart that will show all function calls and how much time was spent in each function execution.
```

> To find slow functions, enter the "Chart" view and search for the last long function in a stack.

Reviewer 3: ```suggestion
To find slow functions, enter the **Chart** view and search for the last long function in a stack.
```

![CPU Flame chart](./media/devtools-chart.png)

> A function's total run time is determined by the runtime of all the functions it calls. You will want to find the one with the longest self runtime, or the last big one before it finishes or splits into smaller calls. For more information on using the DevTools, check out their [official docs](https://developer.chrome.com/docs/devtools/).

Reviewer 3: ```suggestion
A function's total runtime is determined by the runtime of all the functions it calls. You will want to find the function with the longest self runtime, or the last big runtime before it finishes or splits into smaller calls. For more information on using the DevTools, refer to the [official Chrome documentation](https://developer.chrome.com/docs/devtools/).
```

> Here you can see that `sleepBetween` is our culprit. Another quick way of spotting the slowest function in a Worker is to use the Heavy view. This lets you sort by self run time or total run time. Sorting by self run time and ignoring any items in parenthesis will get you your answer.

Reviewer 3: ```suggestion
In the example image above, `sleepBetween`has the longest runtime. Another way to find the slowest function in a Worker is to use the **Heavy** view. With **Heavy** view, you can sort by self runtime or total runtime. Sorting by self runtime and ignoring any items in parenthesis will help you determine the function with the slowest runtime.
```

> In the example image above, `sleepBetween`has the longest runtime. Another way to find the slowest function in a Worker is to use the **Heavy** view. With **Heavy** view, you can sort by self runtime or total runtime. Sorting by self runtime and ignoring any items in parenthesis will help you determine the function with the slowest runtime.

Reviewer 2: ```suggestion
In the example image above, `sleepBetween` has the longest runtime. 

You can also arrive at this conclusion through the **Heavy** view, which allows you to sort by "self runtime" or "total runtime". When you sort by "self runtime" and ignore any items in parenthesis, you can identify the function with the slowest runtime.
```

@deadlypants1973 I don't know if quotation marks are appropriate here, but these are named UI things (right @jkup?)

Author: @lukeed Yes but they are technically "Self Time" and "Total Time" in the devtools.. Updating now!
